Neptune/Services/CatalogApplication.cs

```diff
@@ -0,0 +1,25 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Text;
+
+namespace Neptune.Services
+{
+    interface CatalogApplication : IService
```

Comment:
I'm interested to see how this file turns out. Generally I don't put multiple classes or interfaces in the same file.

Reply:
Everything in here is highly generic, so won't change much with logic changes, and should all change together. What do you think? Should I split it into multiple files?

@ynigoreyes thoughts?

Reply:
I'm questioning why we have to have a class for every service... What if we built out 3 new services that we have not listed. We would need to update this code too. 

I was thinking you can have Get, Post, put, delete and then pass in the service you want to communicate with as an enum. It will reduce our code to maintain by a bunch

Reply:
See: https://docs.microsoft.com/en-us/dotnet/architecture/modern-web-apps-azure/architectural-principles, Single Responsibility Principle.

When applied to OOP each class should have a single responsibility. The message class you've described would both contain the message, and know how to send it. 2 responsibilities that do not evolve together, and 1 (the content) which changes to business logic. This is a coupling between business logic and infrastructure logic.

Another flaw in that plan is the type system should only allow developers to send messages to services which can accept them. The message 'CreateClass' should only be able to be sent the ClassService. That is not possible with the way you've described. We could do a runtime check, but them we would be substituting error 500s and runtime failures for compile time checks, which isn't necessary in a type safe language.

Another reason I've built it like this is the library does not need to contain messages or responses. The service that uses them can control their own contracts IF they want, but common messages can be localized to Neptune.

See https://medium.com/@christophef/sharing-code-between-microservices-b576221820b2 and https://dzone.com/articles/how-to-share-and-sync-code-between-microservices
Shared code can be painful, but I think this is the best way to represent this work in a type safe way. If we want, I think it may be a valuable experience to draft a design document and let the team discuss.

Reply:
I'm kind of confused about what you are building then... I thought you were building out a library that standardized the communication between services. 

I was thinking this is how we will be using it...

```csharp
class Service {
  public Service(IClientFactory clientFactory) {
    this.ClassServiceClient = clientFactory.CreateClient<AvailableServices.ClassService>();
  }

  public async GetAllTheData() {
    // NewClassResponse implements IResponseMessage
    NewClassResponse postResponse = await this.ClassServiceClient.Post<NewClassResponse>('/api/class', new { data: 'Hello World' });
  }
}
```

Reply:
Right, but what you've done here is use an anonymous class. There is no type-safety or agreed upon data contract.

To express this in the Neptune:

```
    interface Class: IService {}
    class Service: ClassService<Class>{
        public async Task<Classroom> CreateClassroom(string name){
            return await this.PostRequest<CreateClasroom, Classroom>(new CreateClasroom(name));
        } 
```
Note:
If you try to send a a CatalogService message to ClassService then the compiler will complain.
If the message contents change in a way that the constructor changes, the compiler will complain.
If you declare that you expect the wrong response, then the compiler will complain.
If the response changes then the compiler will complain.
If a 400/500 is returned then an appropriate exception is thrown.

This is a type safe way to communicate between the services that makes it very difficult to declare nonsense.

Also, the HTTP work is declared in the base class and relies on generic serialization, so there is no boilerplate code to write to send a message.

Reply:
Ahhhh okay okay. That makes sense. You’re right.

And what i mean about boilerplate code is that in order to use Azure Dev Spaces, we need to check for the azds-redirect header that comes into the controllers. I was wondering where you could handle that. Check out the link i sent you